### PR TITLE
NH-3722 - Remove entity mode switching capability

### PR DESCRIPTION
**Pull request by hazzik**
* Ported from HHH-6330 https://github.com/hibernate/hibernate-orm/commit/4a4f636caf9ecc62fe0d230f422ad3eab2517db0
* ISession.GetSession is renamed to ISession.GetChildSession

### CODE REVIEW

File: src/NHibernate/Cfg/SettingsFactory.cs [hunk 1]
```diff
@@ -268,11 +268,6 @@ public Settings BuildSettings(IDictionary<string, string> properties)
 			log.Info("Default flush mode: " + defaultFlushMode);
 			settings.DefaultFlushMode = defaultFlushMode;
 
-			EntityMode defaultEntityMode =
-				EntityModeHelper.Parse(PropertiesHelper.GetString(Environment.DefaultEntityMode, properties, "poco"));
-			log.Info("Default entity-mode: " + defaultEntityMode);
```
Comment (hazzik) [inline src/NHibernate/Cfg/SettingsFactory.cs]: Here, it's better to warn that default entity mode is deprecated rather then silently remove it.

File: src/NHibernate/ISession.cs [hunk 2]
```diff
@@ -932,14 +929,12 @@ public interface ISession : IDisposable
...
-		/// </summary>
-		/// <param name="entityMode">The entity mode to use for the new session.</param>
+		///  <summary>
+		///  Starts a new Session. This secondary Session inherits the connection, transaction,
+		///  and other context information from the primary Session. It doesn't need to be flushed
+		///  or closed by the developer.
+		///  </summary>
 		/// <returns>The new session</returns>
-		ISession GetSession(EntityMode entityMode);
+		ISession GetChildSession();
```
Comment (hazzik) [inline src/NHibernate/ISession.cs]: Need to add this method to 4.1.x and deprecate old one.

File: src/NHibernate/Persister/PersisterFactory.cs [hunk 3]
```diff
@@ -127,18 +115,12 @@ public static class PersisterFactory
...
 		{
 			ConstructorInfo pc;
-			var use4Parameters = false;
 			try
 			{
 				pc = persisterClass.GetConstructor(CollectionPersisterConstructorArgs);
-				if (pc == null)
-				{
-					use4Parameters = true;
-					pc = persisterClass.GetConstructor(CollectionPersisterConstructor2Args);
```
Comment (fredericDelaporte) [inline src/NHibernate/Persister/PersisterFactory.cs]: Could this removal be a breaking change? (Is it possible for NHibernate users to supply custom persister to NHibernate?)

File: src/NHibernate/Type/MetaType.cs [hunk 4]
```diff
@@ -87,11 +86,6 @@ public override bool IsDirty(object old, object current, bool[] checkable, ISess
 			return checkable[0] && IsDirty(old, current, session);
 		}
 
-		public override object FromXMLNode(XmlNode xml, IMapping factory)
-		{
-			return FromXMLString(xml.Value, factory);
-		}
-
 		public object FromXMLString(string xml, IMapping factory)
```
Comment (fredericDelaporte) [inline src/NHibernate/Type/MetaType.cs]: `FromXMLString` looks obsolete, no references found to it. A bit same for `ToXMLString`, only serves for `ToLoggableString` which could directly take `ToXmlString` implementation instead.

File: src/NHibernate/Type/NullableType.cs [hunk 5]
```diff
@@ -343,16 +341,6 @@ public override bool[] ToColumnNullness(object value, IMapping mapping)
...
-		{
-			return FromXMLString(xml.InnerText, factory);
-		}
-
-		public override void SetToXMLNode(XmlNode xml, object value, ISessionFactoryImplementor factory)
-		{
-			xml.InnerText = ToXMLString(value, factory);
-		}
-
 		public string ToXMLString(object value, ISessionFactoryImplementor pc)
```
Comment (fredericDelaporte) [inline src/NHibernate/Type/NullableType.cs]: Same as corresponding methods of `MetaType`: looks obsolete.

**Comment by hazzik** [comment]
>Are we removing the dynamic-map mode altogether? I do not see how we are supposed to activate it since default_entity_mode parameter is removed. On Hibernate side, it looks like they are still supporting it.

The entity mode is supported only on a entity/component level, eg. entity/component has to be defined as dynamic-map, to be able to use it. Please see `DynamicClassFixture`.

Default entity mode was passed only to a SessionImpl constructor, so it's not needed and do not have any effect now.

The drivers for the changes are ComponentMetamodel and EntityMetamodel: replace EntityModeToTuplizerMapping with static ITuplizer and remove all unused EntityModes

>src/UpgradeLog2.htm: that file is probably an unexpected add.
Sure.

File: src/NHibernate/Persister/PersisterFactory.cs (hunk 3 above)
Comment (hazzik) [inline src/NHibernate/Persister/PersisterFactory.cs]: It is, it needs to be deprecated through additional release of 4.x version

**Comment by hazzik** [comment]
>nhibernate-mapping.xsd: still references xml entity mode (line 1435).

I'm not sure how shall we handle xsd schema changes.

**Comment by hazzik** [comment]
>And there is a number of embed-xml attributes which should be removed if they were for the removed embedded xml. (I do not find any reference to those attributes, even in master.)

Because the EntityMode.Xml was never actually implemented

**Comment by hazzik** [comment]
Code issues addressed. I'll update documentation later

**Comment by fredericDelaporte** [comment]
> I'm not sure how shall we handle xsd schema changes.

Me too, but it looks to me your 2 commits have done it right. I guess changing the xsd namespace version would be a hassle for migration, requiring everyone to change it in their existing mappings.

Great work by the way, code base simplification is always welcome.

**Comment by fredericDelaporte** [comment]
On the subject of methods to obsolete, I wonder if this is not a bit late to obsolete them in 4.x for suppressing them completely in 5.
Many users may step over last 4.x versions for upgrading directly to 5.
Maybe should we put them back in 5 but marked obsolete, and remove them only in 6.

**Comment by hazzik** [comment]
I think we can do it for `ISession.GetSession(EntityMode)`. For other methods this does not really make a lot of sense. The thing is, that we are now (in 5.0) in an unique position to reshuffle the API as we need to be able to add new features in the future.

**Pull request merged**